Support for Multiple Devices and Parallelism via MPI

YigitElma opened the pull request:
Initial support for multi-device optimization.

- Updates `set_device` and `desc.backend` for initialization of multiple devices. These are the cause of slightly low coverage (around 90% for patch)
- Adds `_device_id` and `_device` to `_Objective` class (defaults to 0) for making parallelization work with other objectives 
- Uses `mpi4py` for compute and jvp methods of parallel objective functions (there is no separate class but if user passes objectives with different devices and an mpi communicator, parallel strategy will be used.) To make this work a user has to use context manager properly
- Adds `_rank` to `_Objective` that specifies which objective should run on which rank. Default is `None`, and if all sub-objectives have `None`, `ObjectiveFunction` defaults to `np.arange(len(objs))`. There are some limitations for the user. For example, 2 different rank cannot use the same device (maybe it can, but it shouldn't in practice). One tricky part about this is, for multiple node case, same `device_id` can mean different devices. The reason we have `rank` is to be able to distinguish that (MPI.size is equal to total number of devices in all nodes, whereas max(device_id) can be only the number of GPUs connected to a single node)

Things to consider in future PRs,
- Maybe implement a new optimizer that uses distributed matrix operations instead of QR and SVD.
- Find good distributed linear algebra routines for SVD and QR (necessary for speed and low memory usage since the current implementation will use CPU if the Jacobian is big)
  - Something like TSQR implementation via sharding? This should reduce QR of (C*M)xN matrix into C MxN QRs plus a single (C*N)xN QR. By some careful implementation (without storing Q, like in our custom `qr_multiply`), this can be faster.
- Use CUDA-aware MPI (possibly without `mpi4jax`)
- Incorporate sharding if possible?
- Implement MPI version of Augmented Lagrangian optimizers. Currently, they are forced to run on GPU0. This will be tested on a different PR.


TODOS:
- [ ] Make sure the API won't change for the support for constraints

Resolves #1071 (but with `mpi4py`)
Resolves #1601

YigitElma committed:
replicate state vector on all deviecs

YigitElma committed:
allow variable number of gpus, copy some data to every device

YigitElma committed:
not put back to one device for testing

YigitElma committed:
handle num_device=1 case

YigitElma committed:
update

YigitElma committed:
fix typo

YigitElma committed:
fix issue

YigitElma committed:
it was a stupid mistake

YigitElma commented:
Some test scripts,
```python
import nvgpu 
devices = nvgpu.gpu_info()
print([dev["type"] for dev in devices])

from desc import set_device
set_device("gpu", num_device=3)

eq = get("W7-X")
obj = ObjectiveFunction(ForceBalance(eq))
cons = get_fixed_boundary_constraints(eq)
cons = maybe_add_self_consistency(eq, cons)
cons = ObjectiveFunction(cons)
objective = LinearConstraintProjection(objective=obj, constraint=cons)
objective.build()

print(objective.compute_scaled_error(objective.x(eq)).shape)
print(objective.jac_scaled_error(objective.x(eq)).shape)
jac = objective.jac_scaled_error(objective.x(eq))
_, _ = qr(jac, mode="economic") 
%timeit _ = objective.compute_scaled_error(objective.x(eq)).block_until_ready()
%timeit _ = objective.jac_scaled_error(objective.x(eq)).block_until_ready()
%timeit _, _ = qr(jac, mode="economic") 
jax.debug.visualize_array_sharding(jac)
```

YigitElma committed:
I don't know why this was changed

YigitElma committed:
put the copying inside the jitted part

YigitElma committed:
shard A, Z and D too

YigitElma committed:
fix

YigitElma committed:
fix

YigitElma committed:
don't shard A

YigitElma committed:
clean up

YigitElma committed:
shard tangents too

YigitElma committed:
shard v in different way

YigitElma committed:
don't cover set_device for coverage

YigitElma commented:
```python
@jax.jit
def dummy_fun(x):
    return x*jnp.ones(N)

@jax.jit
def dummy_fun2(x):
    return x[0]*jnp.ones(N)

N = 55000
# x = jnp.ones(N)
# v = jnp.eye(N)[:,1]
# f, df = jax.jvp(dummy_fun2, (x,), (v,))

x = 1.
v = 1.
f, df = jax.jvp(dummy_fun, (x,), (v,))
```

dpanici commented:
#763
check for overlap with this one

YigitElma committed:
Merge branch 'master' into yge/multigpu

YigitElma committed:
add getter for parallel force objective

YigitElma committed:
add notebook for testing

YigitElma committed:
build and distribute objectives in getter

YigitElma committed:
maybe use same grid res

YigitElma committed:
add build flag to getter

YigitElma committed:
do not jit the ObjectiveFunction because jax doesn't allow it

YigitElma committed:
move extra stuff

YigitElma committed:
move whole objective on gpu

YigitElma committed:
add pconcat function normal concatenate doesn't accepts arrays from different devices

YigitElma committed:
use more pconcat

YigitElma committed:
test not passing constants

YigitElma committed:
try something

YigitElma committed:
try something

YigitElma committed:
instead replicate eq every device

YigitElma committed:
try something

YigitElma committed:
return replicated eq and use that otherwise outer eq and obj eq are not the same because one is the copy of the other

YigitElma committed:
reorder steps

YigitElma committed:
copy params to device before passing to function

YigitElma committed:
add device_id for forcebalance

YigitElma committed:
update notebook

YigitElma committed:
delete old line

YigitElma committed:
add testing cell

YigitElma committed:
clean up

YigitElma committed:
move params to device for printing too

YigitElma committed:
Merge branch 'master' into yge/multigpu

YigitElma committed:
update notebook to plot grid

Commit:
made it WORK! pass all params on given device, merge arrays on cpu or gpu depending on the size. for very big matrices we move the to cpu to be able call qr or svd. need to use blocked and for loop.

YigitElma committed:
Merge remote-tracking branch 'refs/remotes/origin/yge/multigpu' into yge/multigpu

YigitElma committed:
fix formatting after cluster

YigitElma committed:
Merge branch 'master' into yge/multigpu

YigitElma committed:
fix some problems for testing and docs

YigitElma committed:
ignore multidevice for notebook tests, add additional warnings for gpu, move jit_with_device to backend

YigitElma commented:
This is not yet finalized, but I think this is the main structure I would like to have. Let me know your feedback.

YigitElma committed:
add changelog, fix notebook tests

dpanici commented on desc/objectives/objective_funs.py:
```diff
@@ -436,12 +453,23 @@ def compute_unscaled(self, x, constants=None):
+        else:
+            f = pconcat(
+                [
+                    obj.compute_unscaled(
```
see if we can change this around to make JAX do it asynchronously instead of serially

dpanici replied:
maybe removing this from list comprehension will help?

dpanici committed:
Merge branch 'master' into yge/multigpu

YigitElma committed:
add warning for deriv_mode blocked and moving array to CPU

YigitElma committed:
add option to suppress cpu warning

YigitElma committed:
make upper case

f0uriest commented on desc/__init__.py:
```diff
@@ -58,10 +58,10 @@ def __getattr__(name):
 
 
-def set_device(kind="cpu", gpuid=None):
+def set_device(kind="cpu", gpuid=None, num_device=1):  # noqa : C901
```
though we may also want to allow for multiple CPU devices, useful for testing this stuff. You can pretend a single cpu is multiple devices with
```python
xla_flags = os.getenv("XLA_FLAGS", "")
xla_flags = re.sub(r"--xla_force_host_platform_device_count=\S+", "", xla_flags).split()
os.environ["XLA_FLAGS"] = " ".join(
    [f"--xla_force_host_platform_device_count=8"] + xla_flags
)
```

This would allow us to test this code on CI

YigitElma replied:
Yes, I just saw this on your old PR. I can add a test when I am done with GPU version.

f0uriest commented on desc/backend.py:
```diff
@@ -32,6 +32,15 @@
             jax_config.update("jax_enable_x64", True)
+            if desc_config["num_device"] != 1:
+                # for now, these are not used. Delete them if they are not needed.
+                mesh = jax.make_mesh((desc_config["num_device"],), ("grid"))
```
is this needed? I thought you said you weren't sharding anything?

YigitElma replied:
I will delete them eventually. These were very useful for reducing the amount of code when I was experimenting with it.

f0uriest commented on desc/backend.py:
```diff
@@ -425,6 +443,67 @@ def tangent_solve(g, y):
+        size = jnp.array([x.size for x in arrays])
+        size = jnp.sum(size)
+        if size * 8 / (1024**3) > desc_config["avail_mems"][0]:
+            device = jax.devices("cpu")[0]
```
should at least throw a warning here, or possibly an error.

YigitElma replied:
It now throws a warning that can be suppressed by some configuration parameter. I don't know the proper way of setting flags, but I tried something.

YigitElma replied:
The reason I want to be able to suppress it is that each jacobian evaluation will throw this error and this might get annoying.

f0uriest commented on desc/objectives/_equilibrium.py:
```diff
@@ -61,6 +61,7 @@ def __init__(
         grid=None,
         name="force",
         jac_chunk_size=None,
+        device_id=0,
```
This might be better to pass an instance of `jax.device` or something?

YigitElma replied:
I think from a user standpoint just passing an integer is easier. Otherwise, they have to type `jax.devices("gpu")[0]`

YigitElma replied:
But I can add another attribute `objective._device` which can be jax.device, something like
```python
from desc import config as desc_config
self._device = jax.devices(desc_config['kind'])[self._device_id)]
```
This can make device_put calls cleaner, since currently I assume multiple GPU not CPU.

YigitElma replied:
I implemented above comment.

f0uriest commented on desc/objectives/getters.py:
```diff
@@ -344,3 +344,78 @@ def add_if_multiple(constraints, cls):
     return constraints
+
+
+def get_parallel_forcebalance(
```
not sure if we really want/need this function. I think it's better for now to make the user control objectives and devices more explicitly.

(also I think needing multiple GPUs for a single force objective seems like a rather uncommon use case)

YigitElma replied:
This was mainly a helper function for testing, but I think it could be useful where you have a high res equilibrium you try to optimize using Proximal. Even thought the outer Jacobian is rather small, the ForceBalance part might limit the optimization. So, we can have a this kind of convenience function that is used like,

```python
parallel_force = get_parallel_force(eq, num_device) # assume this returns a tuple 
constraints = (
    *prallel_force,
    FixBoundaryR(eq=eq, modes=R_modes),
    FixBoundaryZ(eq=eq, modes=Z_modes),
    FixPressure(eq=eq),
    FixPsi(eq=eq),
    FixCurrent(eq=eq),
)
```
Basically, this can make grid creation and putting objectives to GPUs automatically.

f0uriest commented on desc/objectives/objective_funs.py:
```diff
@@ -292,19 +297,28 @@ def _unjit(self):
         ----------
         use_jit : bool, optional
             Whether to just-in-time compile the objective and derivatives.
+        use_jit_wrapper : bool, optional
```
does this need to be an argument? shouldn't it be done automatically?

f0uriest commented on desc/objectives/objective_funs.py:
```diff
@@ -292,19 +297,28 @@ def _unjit(self):
         if use_jit is not None:
             self._use_jit = use_jit
+            if use_jit is False:
+                use_jit_wrapper = False
```
need to also set this to something outside of the double if block

YigitElma committed:
clean up set_device

f0uriest reviewed (commented):
some initial thoughts

YigitElma committed:
nuch of clean up

YigitElma committed:
Merge branch 'master' into yge/multigpu

YigitElma committed:
clean up, fix issues

YigitElma committed:
fix set_device config['device'] problem

YigitElma committed:
update notebook and add device_id to all objectives

YigitElma committed:
fix missing docs

YigitElma committed:
initial test for proximal

YigitElma committed:
add obj._device attr for cleaner device_put

YigitElma committed:
jit what you can, use pconcat

YigitElma committed:
fix device jit issue

YigitElma committed:
make _device None for single device cases

YigitElma committed:
ok now it is fixed

YigitElma committed:
Merge branch 'master' into yge/multigpu

YigitElma committed:
implement multicpu, add a test, need to make it work tho

YigitElma committed:
improve test

YigitElma committed:
Merge branch 'master' into yge/multigpu

YigitElma committed:
Merge branch 'master' into yge/multigpu

YigitElma committed:
fix formatting

YigitElma committed:
implement MPI to objective function,update tutorial

YigitElma committed:
add comments

YigitElma committed:
remove redundant lines, fix changelog, update minor stuff

YigitElma committed:
Merge branch 'master' into yge/multigpu

YigitElma committed:
automize obj device placement

YigitElma committed:
add errors for context manager, it must be used by parallel objective and it must be built before entering

YigitElma committed:
update proximal jvp methods, divide constraint and objective part, add batched and blocked options, also implement mpi for proximal

YigitElma committed:
fix for multiple node cases, each process will see only one CPU device, we should still be able to parallelize

YigitElma commented on desc/objectives/utils.py:
```diff
@@ -139,6 +139,7 @@ def factorize_linear_constraints(objective, constraint, x_scale="auto"):  # noqa
     xp = put(xp, unfixed_idx, A_inv @ b)
     xp = put(xp, fixed_idx, ((1 / D) * xp)[fixed_idx])
     # cast to jnp arrays
+    # TODO: might consider sharding these
```
I mean for multiple GPU case, if we use sharding for Z, then `project` and `recover` (which are called a lot) will be faster. It is just a matrix vector product but still...

YigitElma replied:
all sharding options can be tested on a new pr

YigitElma committed:
document more

YigitElma committed:
document more, fix backend print

YigitElma committed:
Merge branch 'master' into yge/multigpu

YigitElma committed:
Merge branch 'master' into yge/multigpu

YigitElma committed:
Merge branch 'master' into yge/multigpu

YigitElma committed:
Merge branch 'master' into yge/multigpu

YigitElma committed:
Merge branch 'master' into yge/multigpu

YigitElma committed:
fixes after merge

YigitElma committed:
first commit for  multiple objective per rank change, need testing

YigitElma committed:
fix blocked mode jac_chunk_size error, now if there is a different sub-objective chunk size will error. Add try/except to get_processor name

YigitElma committed:
fixes after debugging multiple bjective per rank, update tutorials

YigitElma committed:
fix syntax error

YigitElma commented:
@f0uriest @dpanici @rahulgaur104 @ddudt One more review before hackathon would be great.
Updates:
- Each rank can now have multiple objectives. These objectives will be computed in a for-loop as if they are blocked
- For catching some future problems, I set the `ObjectiveFunction._jac_chunk_size` of objectives with multiple objective and sub-objectives have different `jac_chunk_size` to an array (like an error message). If someone ever tries to use it, it will throw an error, and show the error message.

YigitElma committed:
fix the formatting of the notebook

YigitElma committed:
Merge branch 'master' into yge/multigpu

YigitElma committed:
update the memory allocator for GPU

YigitElma committed:
fix the formatting for good

YigitElma committed:
remove repeated jnp.where instead add self._obj_per_rank

rahulgaur104 committed:
Merge branch 'master' into yge/multigpu

rahulgaur104 commented on desc/__init__.py:
```diff
@@ -100,55 +178,62 @@ def set_device(kind="cpu", gpuid=None):
+
+        config["avail_mems"] = [
+            memories[dev["index"]] / 1024 for dev in devices[:num_device]
+        ]  # in GB
```
Is the default value in MB here?

YigitElma replied:
Yeap

rahulgaur104 commented on desc/backend.py:
```diff
@@ -452,11 +472,82 @@ def bodyfun(state):
+        else:
+            device = jax.devices("gpu")[0]
+
+        if mode == "concat":
```
Does this cove all the cases of stacking?
I guess it's only used for objectives or jacobians which can never have more than 2 dimensions?

YigitElma replied:
I think for the parallel stacking applications, it covers all, at least I have never seen a function other than `concatenate`, `hstack` and `vstack`.

rahulgaur104 commented on desc/objectives/getters.py:
```diff
@@ -367,3 +367,63 @@ def add_if_multiple(constraints, cls):
+                f"Number of grids and num_device must be the same! Got "
+                f"{len(grid)=} and {num_device=}."
+            )
+    if eq.L_grid % num_device == 0:
```
This idea of splitting the grid is great only for radially local objectives, right?
If I have an objective where I need the full grid for the compute function this won't work.

YigitElma replied:
Yes, not off the shelf. But maybe one can go around this limitation by changing the objective's compute function. For example, if a quantity requires a special grid, compute it with full grid on each instance (rank/process), then pass it to `compute_fun` through `data`.  But it requires changes to some objectives. I was thinking of using this kind of grid splitting for free-boundary (like obj1->phi=[0,pi/NFP], obj2->phi=[pi/NFP, 2pi/NFP]), we can introduce parallel version or add some keywords to them.

rahulgaur104 commented on desc/objectives/objective_funs.py:
```diff
@@ -267,6 +285,206 @@ def __init__(
+                # inputs to jitted functions must live on the same device. Need to
+                # put xi and vi on the same device as the objective
+                J_rank = []
+                for idx in obj_idx_rank:
```
I am assuming each rank will have it's own value of obj_idx_rank. 
But are all the objectives shared across all the rank?
Because when you do self.objectives[idx], that operation is happening at all ranks,
so each rank must have self.objectives.

YigitElma replied:
Yes, each rank has the built objectives that are on the device they need to be. Only the computation happens on different ranks.

`obj_idx_rank = self._obj_per_rank[self.rank]` is the indices of objectives that belong to this rank, so they are different for each rank.

YigitElma committed:
add CPU model name, they have to be the same anyways

YigitElma committed:
has to be string

YigitElma commented on desc/__init__.py:
```diff
@@ -58,36 +61,114 @@ def __getattr__(name):
+                assert len(jax_cpu) == num_device
+                # These CPUs doesn't have to be the same model, but I think slurm will
+                # always give same model
+                config["devices"] = [f"{cpu_info + " " + str(dev)}" for dev in jax_cpu]
```
```suggestion
                config["devices"] = [f"{cpu_info + ' ' + str(dev)}" for dev in jax_cpu]
```

YigitElma committed:
Update desc/__init__.py

YigitElma committed:
Merge branch 'master' into yge/multigpu

YigitElma commented:
Ok I think instead of

```python
 from desc import _set_cpu_count, set_device
 
 num_device = 2
 # These will be used for diving the single CPU into multiple virtual CPUs
 # such that JAX and XLA thinks there are multiple devices
 # If you have multiple CPUs, this will just use 1 of them, and split it into multiple fake versions
 _set_cpu_count(num_device)
 set_device("cpu", num_device=num_device)
```
I should have this for initializing multiple cpu cases. But just for multiple cpu and 0 gpu case. If there is a GPU current version is fine,

```python
 # to be able to see all CPUs
 # You should start this script using srun
 import jax
 jax.distributed.initialize()
 
 from desc import set_device
 
 num_device = 2
 set_device("cpu", num_device=num_device)
```

YigitElma committed:
make the CPU initialization automatic

rahulgaur104 commented on docs/notebooks/tutorials/mpi-tutorials/mpi-eq-solve.py:
```diff
@@ -0,0 +1,92 @@
+    # useful when you have multiple processes running on multiple nodes and you want
+    # to see which devices are available to each process.
+    if desc_config["kind"] == "gpu":
+        print(
```
This will throw a syntax error because both your outer and inner quotes are ". Make the outer quotes '.

rahulgaur104 replied:
Make the same changes to the tutorial notebook as well.

rahulgaur104 commented on docs/notebooks/tutorials/mpi-tutorials/mpi-proximal.py:
```diff
@@ -0,0 +1,148 @@
+    # to see which devices are available to each process.
+    if desc_config["kind"] == "gpu":
+        print(
+            f"Rank {rank} is running on {jax.local_devices(backend="gpu")} "
```
Same as above.

rahulgaur104 commented on docs/notebooks/tutorials/multi_device.ipynb:
There should be a check for mpi4py installation and maybe a simple test. Because even after installing mpi4py it may not yield the expected behaviour.

Maybe this could be a test
subprocess.call[;'python3 -c "import mpi4py" ']
and use the status to define whther mpi4py exists.

YigitElma replied:
I think this might cause other problems. When you import `mpi4py`, I think it creates the communications, and in some future cases, we might not want that.

Using multiple devices will be definitely a lot trickier than normal version, so I assume the user has proper `mpi4py` installation.

We can consider adding that during hackathon though

rahulgaur104 commented on docs/notebooks/tutorials/mpi-tutorials/mpi-eq-solve.py:
```diff
@@ -0,0 +1,92 @@
+from desc import _set_cpu_count, set_device
+
+# ====== Using CPUs ======
+num_device = 4
```
when I set num_device = 1, I get the following error

```python3

mpirun -n 1 python3 -u mpi-eq-solve.py 
====== TOTAL OF 1 RANKS ======
Rank 0 is running on [CpuDevice(id=0)]

====== BACKEND INFO ======
DESC version=0.14.2+133.g50e5239ed.dirty.
Using JAX backend: jax version=0.4.31, jaxlib version=0.4.31, dtype=float64.
CPU Info:  AMD Ryzen 7 7735U with Radeon Graphics CPU with 21.98 GB available memory


Starting timer!
Building objective: force
Precomputing transforms
------------------------------------------------------------
Rank 0 will run objective(s): ['ForceBalance']
------------------------------------------------------------
Building objective: lcfs R
Building objective: lcfs Z
Building objective: fixed Psi
Building objective: fixed pressure
Building objective: fixed iota
Building objective: fixed sheet current
Building objective: self_consistency R
Building objective: self_consistency Z
Building objective: lambda gauge
Building objective: axis R self consistency
Building objective: axis Z self consistency
Timer: Objective build = 1.03 sec
Timer: LinearConstraintProjection build = 4.17 sec
Number of parameters: 551
Number of objectives: 8424
Timer: Initializing the optimization = 5.24 sec

Starting optimization
Using method: lsq-exact
Traceback (most recent call last):
  File "/home/rgaur/DESC/docs/notebooks/tutorials/mpi-tutorials/mpi-eq-solve.py", line 84, in <module>
    eq.solve(
  File "/home/rgaur/DESC/desc/equilibrium/equilibrium.py", line 2254, in solve
    things, result = optimizer.optimize(
                     ^^^^^^^^^^^^^^^^^^^
  File "/home/rgaur/DESC/desc/optimize/optimizer.py", line 278, in optimize
    result = optimizers[method]["fun"](
             ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rgaur/DESC/desc/optimize/_desc_wrappers.py", line 270, in _optimize_desc_least_squares
    result = lsqtr(
             ^^^^^^
  File "/home/rgaur/DESC/desc/optimize/least_squares.py", line 178, in lsqtr
    f = fun(x, *args)
        ^^^^^^^^^^^^^
  File "/home/rgaur/DESC/desc/optimize/_constraint_wrappers.py", line 314, in compute_scaled_error
    f = self._objective.compute_scaled_error(x, constants)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: Cannot interpret value of type <class 'mpi4py.MPI.Intracomm'> as an abstract array; it does not have a dtype attribute
--------------------
For simplicity, JAX has removed its internal frames from the traceback of the following exception. Set JAX_TRACEBACK_FILTERING=off to include these.
--------------------------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
--------------------------------------------------------------------------
--------------------------------------------------------------------------
mpirun detected that one or more processes exited with non-zero status, thus causing
the job to be terminated. The first process to do so was:

  Process name: [[38784,1],0]
  Exit code:    1
--------------------------------------------------------------------------

```

YigitElma replied:
I added an error for this case.

rahulgaur104 commented:
So I only compared the mpi-eq-solve script timings with 2 and 4 processes. With 2 processes it takes ~ 31 s but with 4 it takes ~ 180s. This is probably because a lot of time is being spent on gather compute_scaled_error data.

Maybe this will change when we use multiple GPUs.

YigitElma commented:
> So I only compared the mpi-eq-solve script timings with 2 and 4 processes. With 2 processes it takes ~ 31 s but with 4 it takes ~ 180s. This is probably because a lot of time is being spent on gather compute_scaled_error data.
> 
> Maybe this is happening because the grid cannot be divided across 4 CPU and not the communication.
> 

What are the number of parameters and number of objectives for both cases?

If it is the same, yes probably the transfer overhead is high. If they are not, then it is peobably the QR decomposition difference. Because QR is still happening on single device.

YigitElma committed:
add error for 1 device case with MPI, fix syntax

YigitElma committed:
fixes for multiple CPU data transfer

YigitElma committed:
don't try to initialize twice

YigitElma committed:
test

YigitElma committed:
test

YigitElma committed:
test

YigitElma committed:
add distributed initilize back, update error for empty ranks

YigitElma committed:
distributed initialize will require bunch of additional code since it tries to put each array to a single CPU that is remote to some other processes and hence error out. This removes distributed initialize. Updates some error cases

YigitElma commented:
Annoying thing about `jax.distributed.initialize()`:

It allows you to see other devices that are not local to the process. However, for some reason, makes the default device, the CPU of the rank 0 process. So, rank 0 is fine, but other processes are actually using a remote CPU that is not local to them. This throws an error `jaxlib.xla_extension.XlaRuntimeError: INVALID_ARGUMENT: Cannot copy array to non-addressable device: BasicDeviceList([CpuDevice(id=0)])` which is triggered by `jnp.array` and `jnp.asarray` operations. Unless someone is willing to add the device flag to each of these functions (and make sure that future codes also have that flag), I would recommend not starting distributed initialize.

YigitElma committed:
fix error check

YigitElma committed:
gather cpu infor ranks by MPI

YigitElma committed:
add notes to the backend print

YigitElma commented:
@rahulgaur104 I made some updates. Here is the lean-up output I get for 4 CPUs equilibrium solve (you need to comment out `_set_cpu_count()` line)

```bash
====== TOTAL OF 4 RANKS ======
Rank 0 can see [CpuDevice(id=0)]

====== BACKEND INFO ======
DESC version=0.14.2+144.ge93d0c937.dirty.
Using JAX backend: jax version=0.6.0, jaxlib version=0.6.0, dtype=float64.
Using 4 CPUs with 1376.06 GB total available memory:
	 CPU : 0  AMD EPYC 9654 96-Core Processor
	 CPU : 1  Intel(R) Xeon(R) Gold 6242R CPU @ 3.10GHz
	 CPU : 2  Intel(R) Xeon(R) Gold 6242R CPU @ 3.10GHz
	 CPU : 3  Intel(R) Xeon(R) Gold 6242R CPU @ 3.10GHz

Note: The backend information assumes that the user has 1 process per CPU (node). Using multiple processes per CPU (node) is not the most efficient way to use MPI with purely CPUs.

------------------------------------------------------------
Rank 0 will run objective(s): ['ForceBalance']
Rank 1 will run objective(s): ['ForceBalance']
Rank 2 will run objective(s): ['ForceBalance']
Rank 3 will run objective(s): ['ForceBalance']
------------------------------------------------------------
Timer: Objective build = 1.07 sec
Building objective: force
Timer: LinearConstraintProjection build = 3.86 sec
Number of parameters: 551
Number of objectives: 8424
Timer: Initializing the optimization = 4.97 sec

Starting optimization
Using method: lsq-exact
   Iteration     Total nfev        Cost      Cost reduction    Step norm     Optimality   
       0              1          2.500e+00                                    1.228e+00   
       1              2          8.226e-01      1.678e+00      2.256e-01      5.198e-01   
       2              3          2.473e-02      7.978e-01      1.993e-01      7.096e-02   
       3              5          3.451e-03      2.128e-02      8.908e-02      3.526e-02   
Warning: Maximum number of iterations has been exceeded.
         Current function value: 3.451e-03
         Total delta_x: 2.505e-01
         Iterations: 3
         Function evaluations: 5
         Jacobian evaluations: 4
Timer: Solution time = 17.8 sec
Timer: Avg time per step = 4.46 sec
================================================================================================
                                                                 Start  -->   End
Total (sum of squares):                                      2.500e+00  -->   3.451e-03, 
Maximum absolute Force error:                                6.794e+04  -->   6.655e+03 (N)
Minimum absolute Force error:                                1.059e-10  -->   1.232e-10 (N)
Average absolute Force error:                                1.304e+04  -->   1.696e+03 (N)
Maximum absolute Force error:                                5.464e-03  -->   5.352e-04 (normalized)
Minimum absolute Force error:                                8.514e-18  -->   9.912e-18 (normalized)
Average absolute Force error:                                1.049e-03  -->   1.364e-04 (normalized)
Maximum absolute Force error:                                2.032e+05  -->   1.704e+04 (N)
Minimum absolute Force error:                                1.502e-10  -->   1.535e-10 (N)
Average absolute Force error:                                4.494e+04  -->   2.436e+03 (N)
Maximum absolute Force error:                                1.634e-02  -->   1.370e-03 (normalized)
Minimum absolute Force error:                                1.208e-17  -->   1.235e-17 (normalized)
Average absolute Force error:                                3.614e-03  -->   1.959e-04 (normalized)
Maximum absolute Force error:                                6.807e+05  -->   2.238e+04 (N)
Minimum absolute Force error:                                8.896e-11  -->   6.677e-11 (N)
Average absolute Force error:                                7.775e+04  -->   2.548e+03 (N)
Maximum absolute Force error:                                5.474e-02  -->   1.800e-03 (normalized)
Minimum absolute Force error:                                7.155e-18  -->   5.370e-18 (normalized)
Average absolute Force error:                                6.253e-03  -->   2.049e-04 (normalized)
Maximum absolute Force error:                                1.149e+07  -->   1.001e+06 (N)
Minimum absolute Force error:                                3.304e-12  -->   6.674e-13 (N)
Average absolute Force error:                                1.493e+05  -->   6.575e+03 (N)
Maximum absolute Force error:                                9.238e-01  -->   8.054e-02 (normalized)
Minimum absolute Force error:                                2.657e-19  -->   5.368e-20 (normalized)
Average absolute Force error:                                1.200e-02  -->   5.288e-04 (normalized)
R boundary error:                                            0.000e+00  -->   0.000e+00 (m)
Z boundary error:                                            0.000e+00  -->   0.000e+00 (m)
Fixed Psi error:                                             0.000e+00  -->   0.000e+00 (Wb)
Fixed pressure profile error:                                0.000e+00  -->   0.000e+00 (Pa)
Fixed iota profile error:                                    0.000e+00  -->   0.000e+00 (dimensionless)
Fixed sheet current error:                                   0.000e+00  -->   0.000e+00 (~)
================================================================================================

Rank 3 STOPPING
Rank 1 STOPPING
Rank 2 STOPPING
```

YigitElma commented:
This is for 2 CPUs,

```bash
====== TOTAL OF 2 RANKS ======
Rank 0 can see [CpuDevice(id=0)]

====== BACKEND INFO ======
DESC version=0.14.2+144.ge93d0c937.dirty.
Using JAX backend: jax version=0.6.0, jaxlib version=0.6.0, dtype=float64.
Using 2 CPUs with 1378.87 GB total available memory:
	 CPU : 0  AMD EPYC 9654 96-Core Processor
	 CPU : 1  Intel(R) Xeon(R) Gold 6242R CPU @ 3.10GHz

Note: The backend information assumes that the user has 1 process per CPU (node). Using multiple processes per CPU (node) is not the most efficient way to use MPI with purely CPUs.

------------------------------------------------------------
Rank 0 will run objective(s): ['ForceBalance']
Rank 1 will run objective(s): ['ForceBalance']
------------------------------------------------------------
Timer: Objective build = 1.04 sec
Timer: LinearConstraintProjection build = 3.70 sec
Number of parameters: 551
Number of objectives: 8424
Timer: Initializing the optimization = 4.78 sec

Starting optimization
Using method: lsq-exact
   Iteration     Total nfev        Cost      Cost reduction    Step norm     Optimality   
       0              1          8.248e-01                                    6.170e-01   
       1              2          3.863e-01      4.385e-01      4.568e-01      2.795e-01   
       2              3          1.176e-01      2.687e-01      2.927e-01      1.966e-01   
       3              4          8.549e-02      3.208e-02      3.424e-01      1.955e-01   
Warning: Maximum number of iterations has been exceeded.
         Current function value: 8.549e-02
         Total delta_x: 5.986e-01
         Iterations: 3
         Function evaluations: 4
         Jacobian evaluations: 4
Timer: Solution time = 17.9 sec
Timer: Avg time per step = 4.48 sec
================================================================================================
                                                                 Start  -->   End
Total (sum of squares):                                      8.248e-01  -->   8.549e-02, 
Maximum absolute Force error:                                2.032e+05  -->   7.521e+04 (N)
Minimum absolute Force error:                                1.059e-10  -->   1.375e-10 (N)
Average absolute Force error:                                4.102e+04  -->   1.600e+04 (N)
Maximum absolute Force error:                                1.634e-02  -->   6.049e-03 (normalized)
Minimum absolute Force error:                                8.514e-18  -->   1.106e-17 (normalized)
Average absolute Force error:                                3.299e-03  -->   1.287e-03 (normalized)
Maximum absolute Force error:                                1.149e+07  -->   1.580e+06 (N)
Minimum absolute Force error:                                3.304e-12  -->   1.282e-12 (N)
Average absolute Force error:                                1.029e+05  -->   2.220e+04 (N)
Maximum absolute Force error:                                9.238e-01  -->   1.270e-01 (normalized)
Minimum absolute Force error:                                2.657e-19  -->   1.031e-19 (normalized)
Average absolute Force error:                                8.279e-03  -->   1.785e-03 (normalized)
R boundary error:                                            0.000e+00  -->   0.000e+00 (m)
Z boundary error:                                            0.000e+00  -->   0.000e+00 (m)
Fixed Psi error:                                             0.000e+00  -->   0.000e+00 (Wb)
Fixed pressure profile error:                                0.000e+00  -->   0.000e+00 (Pa)
Fixed iota profile error:                                    0.000e+00  -->   0.000e+00 (dimensionless)
Fixed sheet current error:                                   0.000e+00  -->   0.000e+00 (~)
================================================================================================
Rank 1 STOPPING

```

YigitElma committed:
Merge branch 'master' into yge/multigpu

dpanici committed:
Merge branch 'master' into yge/multigpu

YigitElma committed:
Merge branch 'master' into yge/multigpu

YigitElma committed:
fix the jax.local_devices error for gpu backend

YigitElma committed:
fix the scripts os path

YigitElma committed:
add the script with nvtx

YigitElma committed:
add force balance nvtx script, make the nvtx scripts more demanding

dpanici commented:
Note that when running on cluster using a command like `srun -n 2 python mpi-eq-solve.py`, the `-n X` part seems like it must match what you requested (and what your number of devices is in the code) in order to avoid errors/hanging code (in my brief experience).

Perhaps we could use some env flags to avoid having to include this in the script, like using `import os; num_devices = os.environ["SLURM_CPUS_PER_TASK"]`?

YigitElma committed:
for hackathon purposes add nvtx flags, add 2 node script

YigitElma committed:
solve concatenation bug for scalar and non-sclaar objective grad/jacobian

YigitElma commented:
> Note that when running on cluster using a command like `srun -n 2 python mpi-eq-solve.py`, the `-n X` part seems like it must match what you requested (and what your number of devices is in the script itself) in order to avoid errors/hanging code (in my brief experience).
> 
> Perhaps we could use some env flags to avoid having to include this in the script, like using `import os; num_devices = os.environ["SLURM_CPUS_PER_TASK"]`?
> 
> EDIT: I think this actually was due to slow compile times so I just thought it was hanging. I still think the second point is important

Yes, the script and `srun -n X ... ` must match but you can use less than or equal to number of processes you previously requested during slurm file. For example, requesting 4 and using only 2 (with proper script) works fine. 

Other combinations, like not using all the requested ranks (supplied by srun) will throw an error during the ObjectiFunction phase, saying that you should put at least one objective to all ranks.

dpanici commented:
- utility to split up a single objective and single grid into multiple obj and multiple grids for parallel distribution

YigitElma committed:
Merge branch 'master' into yge/multigpu

dpanici committed:
change hstack to vstack to fix concat error

YigitElma committed:
add more and more nvtx flags for stages

YigitElma committed:
fix some of the annotations

YigitElma committed:
add annotate to unpacking

YigitElma committed:
jit whatever you can, works on cpu, hopefull will on GPUs. unpack_state and per process computations (for loops) are jitted, should be faster?

YigitElma committed:
jit the functions on the proper device

YigitElma committed:
move the jitted function out of self to prevent recompilation for compute methods

YigitElma committed:
add one more annotate

YigitElma committed:
fix the jvp jit

YigitElma committed:
remove numpy calls to use cuda aware mpi

YigitElma committed:
jit the proximal too, re add the numpy calls until we figure out how to install cuda aware mpi

YigitElma committed:
add annotations and use jit for proximal everywhere

YigitElma committed:
add more flags, form lists outside

YigitElma committed:
fix block_until_ready

YigitElma committed:
no mpi benchmark case

YigitElma committed:
Merge branch 'master' into yge/multigpu

YigitElma committed:
increase maxiter

YigitElma committed:
remove constants from args

YigitElma committed:
Merge branch 'master' into yge/multigpu

YigitElma committed:
remove jit_with_device, remove some debugging

YigitElma commented:
Some profiles using JAX 0.6.2 (MPI without CUDA) for `eq-solve` example 2 processes
![image](https://github.com/user-attachments/assets/5b5df754-6c21-44c2-887f-d5875b78d6a9)
Close up on `compute`,
![image](https://github.com/user-attachments/assets/cc99ec7c-9947-4e3d-a0d4-bfc920fc8939)
Close up on `jacobian`,
![image](https://github.com/user-attachments/assets/11ad6578-5296-47a4-84b8-110f8e13d516)


Maybe the MPI communication overhead can be reduced by CUDA-aware MPI (we didn't spend much time on that during hackathon). But still, there will be some overhead anyway due to shear array sizes (i.e. say portion of the Jacobian is 1GB, then it will take 100ms with 10GB/s NVLink connection at best, tho I am not sure about the exact speeds))

YigitElma committed:
update changelog

YigitElma committed:
update tutorials

YigitElma committed:
Merge branch 'master' into yge/multigpu

dpanici commented:
@dpanici find use case where it speeds things up (auglag to avoid eq solves)

dpanici commented:
Dependency stuff: add MPI as optional dependency (in pip too), would also be good to point to instructions for openMPI maybe
make it so DESC works fine without MPI installed.

YigitElma committed:
Merge branch 'master' into yge/multigpu

YigitElma committed:
clean debugging code

YigitElma committed:
remove debug utility function get_parallel_forcebalance, it wasn't meant for final use

YigitElma committed:
Merge branch 'master' into yge/multigpu

YigitElma commented on desc/objectives/objective_funs.py:
```diff
@@ -1103,6 +1476,9 @@ class _Objective(IOAble, ABC):
         "_scalar",
         "_units",
     ]
+    # _device is of type 'jaxlib.xla_extension.Device' which cannot be an argument
+    # to a jitted function.
+    _static_attrs = ["_device"]
```
```suggestion
        "_scalar",
        "_units",
        "_device",
    ]
```

YigitElma committed:
Update desc/objectives/objective_funs.py

YigitElma committed:
fix static args

YigitElma committed:
add initial attempt for mpi tests

YigitElma committed:
I don't know why this worked??? add some extra tests

YigitElma committed:
Merge branch 'master' into yge/multigpu

YigitElma committed:
clean up redundant jits

YigitElma committed:
fix the failing test, had to build it before and used correct method name

YigitElma committed:
add back the space

YigitElma committed:
Merge branch 'master' into yge/multigpu

YigitElma committed:
I don't know why I need to add _static_attrs to _static_attrs only for this PR but it solves the issue

YigitElma committed:
make the mpi test work

YigitElma committed:
remove redundant function

YigitElma committed:
add more tests, remove some no-coverage commands

YigitElma committed:
add compute and derivatove tests, some clean-up

YigitElma committed:
Merge remote-tracking branch 'origin/master' into yge/multigpu

YigitElma committed:
fix the stupid bug, another reason why #1728 is necessary

YigitElma committed:
add tests, check errors/warnings

YigitElma committed:
try to re-use code as much as possible, add oversubscribe to prevent MPI test fail randomly

YigitElma committed:
fix tests

YigitElma committed:
try to fix test

YigitElma committed:
try to fix test

YigitElma committed:
Merge branch 'master' into yge/multigpu

YigitElma committed:
write coverage to separate files to prevent error

YigitElma committed:
update changelog

YigitElma commented on desc/objectives/objective_funs.py:
```diff
@@ -314,30 +507,52 @@ def _unjit(self):
+                        f"{objective._device_id}"
+                    )
+                objective = jax.device_put(objective, objective._device)
+                objective._things = obj_things
```
Explain this with a comment. 

We do this to prevent different copies of the same thing in different devices. This is necessary for finding proper unique things.

YigitElma commented on desc/objectives/objective_funs.py:
```diff
@@ -475,12 +720,21 @@ def compute_unscaled(self, x, constants=None):
-                for par, obj, const in zip(params, self.objectives, constants)
-            ]
-        )
+        if not self._is_mpi:
```
We can even have a single _compute_op function that takes op and couple other stuff and do the computation which is kind of repeat now

YigitElma commented on desc/objectives/objective_funs.py:
```diff
@@ -504,12 +758,17 @@ def compute_scaled(self, x, constants=None):
-                for par, obj, const in zip(params, self.objectives, constants)
-            ]
-        )
+        if not self._is_mpi:
```
See above

YigitElma commented on desc/objectives/objective_funs.py:
```diff
@@ -533,12 +792,21 @@ def compute_scaled_error(self, x, constants=None):
-                for par, obj, const in zip(params, self.objectives, constants)
-            ]
-        )
+        if not self._is_mpi:
```
See above

YigitElma commented on desc/objectives/objective_funs.py:
```diff
@@ -695,18 +969,28 @@ def grad(self, x, constants=None):
+            )
+        else:
+            raise NotImplementedError(
+                "Gradient computation is not implemented for MPI ObjectiveFunction."
```
This is probably easy to implement via f.T@Jacobian where both f and Jacobian can be computed in parallel

YigitElma commented on desc/objectives/objective_funs.py:
```diff
@@ -1375,6 +1701,8 @@ def jac_unscaled(self, *args, **kwargs):
         )(*args, **kwargs)
 
     def _jvp(self, v, x, constants=None, op="scaled"):
+        if constants is None:
```
Need this in master too

YigitElma commented on desc/optimize/_constraint_wrappers.py:
```diff
@@ -1198,23 +1199,34 @@ def _jvp(self, v, x, constants=None, op="scaled_error"):
+            # TODO: implement parallel constraint for ProximalProjection
+            raise NotImplementedError(
+                "Parallel constraint for ProximalProjection not implemented yet. Note: "
+                "This would require putting workers into a second infinite loop which "
```
Leave the Note: part as comment, user doesnt need to know all of this

YigitElma commented on docs/notebooks/tutorials/mpi-tutorials/mpi-eq-solve.py:
```diff
@@ -0,0 +1,116 @@
+
+    # this context manager will put the workers in a loop to listen to the master
+    # to compute the objective function and its derivatives
+    with obj as obj:
```
Do we need 'as obj' part?

YigitElma commented on docs/notebooks/tutorials/mpi-tutorials/mpi-proximal.py:
```diff
@@ -0,0 +1,160 @@
+
+    # this context manager will put the workers in a loop to listen to the master
+    # to compute the objective function and its derivatives
+    with objective as objective:
```
Same as above

YigitElma commented on docs/notebooks/tutorials/multi_device.ipynb:
```diff
@@ -0,0 +1,793 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
```
Is there a better way to take the content of files here and display as code? It is very tedious to keep them in sync

YigitElma commented on docs/notebooks/tutorials/multi_device.ipynb:
```diff
@@ -0,0 +1,793 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
```
Same as above

YigitElma committed:
address my own comments

YigitElma commented on desc/objectives/objective_funs.py:
```diff
@@ -283,6 +302,180 @@ def __init__(
+                if rank_per_objective is not None
+                else np.arange(len(objectives))
+            )
+            errorif(
```
One can probably find more efficient way to validate inputs, I kind of added more errors as I came up with them. Appreciate any feedback!

YigitElma committed:
remove redundant change to make reviewing easier

YigitElma commented:
I am about to make some changes that might break some code!

The main goal is to use capitalized versions of `bcast` and `gather` to reduce overhead. I have read that `bcast` and `gather` (lower case versions) are meant for general Python objects, so they use pickling. Capitalized versions are told to be more efficient.

YigitElma committed:
use Bcast for arrays instead of bcast which uses pickling and allegdly has overhead

YigitElma committed:
seems like FixParameters is changing dim_x? The tests that use FixParameters were failing when dim_x_splits was static, reverting (also wasn't needed anymore)

YigitElma committed:
Merge remote-tracking branch 'origin/master' into yge/multigpu

YigitElma committed:
update changelog and docs

YigitElma committed:
use Gatherv for passing Jacobian and compute data

YigitElma committed:
resolve the scrambled array issue, add test for Proximal derivatives

YigitElma committed:
remove pconcat since it is not needed anymore

YigitElma committed:
fix dimensional issue, add linear constraint projection test too

YigitElma committed:
trying to fix CUDA aware MPI

YigitElma committed:
trying to fix CUDA aware MPI, add block until ready

YigitElma committed:
make the device transfer based on desc_config[mpi-cuda] which is False until we make it CUDA-aware compatible

YigitElma committed:
add comment

YigitElma committed:
remove some block until ready s

YigitElma committed:
Merge branch 'master' into yge/multigpu

YigitElma committed:
fix the unintended copy by returning the new stuff and overwriting the original by the new

dpanici committed:
Merge branch 'master' into yge/multigpu

YigitElma committed:
Merge branch 'master' into yge/multigpu

YigitElma committed:
Merge branch 'master' into yge/multigpu

YigitElma committed:
Merge branch 'master' into yge/multigpu

YigitElma committed:
Merge branch 'master' into yge/multigpu

YigitElma committed:
Merge branch 'master' into yge/multigpu

YigitElma committed:
gather arrays on CPU, need to add a size check later

YigitElma committed:
add size check for the transfer

YigitElma committed:
resurrect memory warning

YigitElma commented on desc/backend.py:
```diff
@@ -521,6 +554,69 @@ def bodyfun(state):
+                    "`from desc import config as desc_config` \n"
+                    "`desc_config['SUPPRESS_GPU_MEMORY_WARNING'] = True`"
+                )
+            return jnp.asarray(arr, device=jax.devices("cpu")[0])
```
Thinking about this again, I don't think this is enough to conduct  trust region operations on CPU. I probably need to add `@execute_on_cpu` to those functions, optionally. So, things will get slightly messy.

YigitElma committed:
fix the proximal grad problem, must implement proper vjp later

YigitElma committed:
Merge remote-tracking branch 'origin/master' into yge/multigpu

YigitElma committed:
add some possible fix to memory issue, not pleasant yet

YigitElma committed:
ok, since J is commited to a device, computation should follow where J is, see jax discussion #34942

YigitElma committed:
reduce xtol, CI runner has a different processor which fails, cannot reproduce locally

ddudt committed:
Merge branch 'master' into yge/multigpu

ddudt committed:
Merge branch 'master' into yge/multigpu

ddudt committed:
Merge branch 'master' into yge/multigpu

ddudt commented:
> @rahulgaur104 @ddudt I could make this work on NERSC Perlmutter with the following steps,

I tested this today and it worked. Thanks!

YigitElma committed:
Merge branch 'master' into yge/multigpu

ddudt committed:
Merge branch 'master' into yge/multigpu

daniel-dudt committed:
add device_id to PlasmaCoilSetMinDistance

ddudt commented on desc/objectives/_coils.py:
```diff
@@ -1391,6 +1407,7 @@ def __init__(
         use_softmin=False,
         softmin_alpha=1.0,
         dist_chunk_size=None,
+        device_id=0,
```
I noticed `device_id` was missing as an argument for `PlasmaCoilSetMinDistance` so I added it here. We should test or check that this works for every objective in case it is missing anywhere else.

YigitElma replied:
Oh thanks, I added these to all objectives a year ago, so newer objectives might be missing. For general stuff like chunk size, device id, target etc, we might actually pass them as kwargs, but a low priority

ddudt committed:
Merge branch 'master' into yge/multigpu

YigitElma committed:
resolve the todo, do not allow linear constraints to have different device, update changelog

YigitElma committed:
remove redundant changes

YigitElma committed:
Merge branch 'master' into yge/multigpu

YigitElma committed:
Merge branch 'master' into yge/multigpu

YigitElma committed:
formatting after merge conflict

YigitElma committed:
fix test

YigitElma committed:
Merge remote-tracking branch 'origin/master' into yge/multigpu

YigitElma committed:
update pytest marks

YigitElma committed:
remove some unnecessary changes

ddudt commented on desc/__init__.py:
```diff
@@ -58,37 +61,134 @@ def __getattr__(name):
+    )
+
+
+def set_device(kind="cpu", gpuid=None, num_device=1, mpi=None):  # noqa: C901
```
[![todo(blocking)](https://img.shields.io/badge/todo-blocking-374151?labelColor=E879F9)](https://pullpo.io/cc?l=todo&d=blocking) 
`mpi` argument is not documented. It looks like it is only used if `kind="cpu"` and not for the GPU?

YigitElma replied:
Yes, it is only used for CPU backend. I updated the docstring. It is actually not too useful, but for users without GPU access can give better information. Supplying `mpi` can only change the output of `print_backend_info()`, otherwise it doesn't affect resource usage.

ddudt commented on .github/workflows/notebook_tests.yml:
[![question(non-blocking)](https://img.shields.io/badge/question-non--blocking-9CA3AF?labelColor=8B5CF6)](https://pullpo.io/cc?l=question&d=non-blocking) 
Is this a necessary change to increase the indentation? It looks like the only real change is to add the new notebook

YigitElma replied:
I don't think so. I just remember adding the notebook, maybe black formatted it later?

ddudt commented on docs/notebooks/tutorials/mpi-tutorials/mpi-eq-solve.py:
```diff
@@ -0,0 +1,116 @@
+kind = "cpu"  # or "gpu"
+num_device = 2
+# ====== Using CPUs ======
+# These will be used for diving the single CPU into multiple virtual CPUs
```
diving -> dividing

ddudt commented on desc/objectives/objective_funs.py:
```diff
@@ -348,6 +358,12 @@ class ObjectiveFunction(IOAble):
         to manually choose a chunk_size if an OOM error is experienced in this case.
+    mpi : MPI object, optional
+        MPI communicator. Required when using multiple devices.
+    rank_per_objective : ndarray, optional
```
[![suggestion(if-minor)](https://img.shields.io/badge/suggestion-if--minor-14B8A6?labelColor=3B82F6)](https://pullpo.io/cc?l=suggestion&d=if-minor) 
Could we allow this to also be a list? We could cast the user input to an array if they give a list/tuple.

ddudt commented on docs/notebooks/tutorials/multi_device.ipynb:
[![suggestion](https://img.shields.io/badge/suggestion-3B82F6)](https://pullpo.io/cc?l=suggestion) 
Note about the last part of this notebook, where you show an example of running 6 objectives across 2 nodes with 3 GPUs per node: 

> By default, the initializer will assign different ranks for each sub-objective.

It is not obvious from the example how it specifies which objectives are assigned to node 1 vs node 2. This is done with `rank_per_objective`, assuming the default value of `np.arange(len(objectives))`. I recommend showing this `rank_per_objective` input in the example, to clarify that is how it is using both nodes.

ddudt commented on desc/objectives/objective_funs.py:
```diff
@@ -404,6 +423,211 @@ def __init__(
+                f"ranks {self._rank_per_objective} and device ids {device_ids} are "
+                "not compatible.",
+            )
+            # TODO: should this throw an Error?
```
[![thought](https://img.shields.io/badge/thought-6B7280)](https://pullpo.io/cc?l=thought) 
I think this is fine as a warning, as long as it does not break the code and it will still run. Some nodes have 8 GPUs and maybe the user only has a few different objectives, so it is inefficient but we should allow it.

ddudt commented on tests/test_multidevice.py:
```diff
@@ -0,0 +1,499 @@
+                verbose=3,
+            )
+            f1 = objective.compute_scalar(objective.x(eq))
+            assert f1 < f0
```
[![suggestion(if-minor)](https://img.shields.io/badge/suggestion-if--minor-14B8A6?labelColor=3B82F6)](https://pullpo.io/cc?l=suggestion&d=if-minor) 
This test is only checking that the optimization runs and decreases the cost. We could also run the same optimization on a single device and check that both optimizations give the same result. It would take longer to run but be a more robust test.

YigitElma replied:
I can write the same optimization with a single device, again in blocked mode, and run it before this. Splitting the objectives into multiple sub-objectives is necessary in the single device case because otherwise, the ordering of the grid nodes differs and causes different numerics. That is why I was using this strategy for the other tests above.

ddudt reviewed (commented):
[![praise](https://img.shields.io/badge/praise-28A745)](https://pullpo.io/cc?l=praise) 
I've tested this PR on Perlmutter with a single node and it works very well! I have a few minor comments, otherwise I will approve. I will also try to test with multiple nodes.

daniel-dudt committed:
clean up comments

ddudt committed:
Merge branch 'master' into yge/multigpu

ddudt commented:
I tested this on multiple nodes and it works as expected. It wasn't very slow with 2 nodes on Perlmutter.

YigitElma committed:
remove indent

YigitElma committed:
allow array-like inputs for rank_per_objective

YigitElma committed:
update tutorial

YigitElma committed:
update the test to compare actual optimization results

YigitElma committed:
update the tutorial

YigitElma committed:
update docstring for set_device

YigitElma committed:
Merge branch 'master' into yge/multigpu

YigitElma commented on desc/objectives/objective_funs.py:
```diff
@@ -1565,6 +1951,8 @@ def jac_unscaled(self, *args, **kwargs):
 
     def _jvp(self, v, x, constants=None, op="scaled"):
+        if constants is None:
+            constants = self.constants
```
```suggestion
            constants = self._get_deprecated_constants(constants)
```

YigitElma commented on desc/objectives/objective_funs.py:
```diff
@@ -2002,3 +2390,48 @@ def __call__(self, things):
+        else:
+            J_rank.append(
+                getattr(obj, "jvp_" + op)(
+                    [_vi for _vi in v[idx]], x[idx], constants=obj.constants
```
```suggestion
                    [_vi for _vi in v[idx]], x[idx], constants=None
```

YigitElma commented on desc/objectives/objective_funs.py:
```diff
@@ -2002,3 +2390,48 @@ def __call__(self, things):
+    """Compute the Jacobian-vector product on each process."""
+    return jnp.hstack(
+        [
+            getattr(obj, op)(v[idx], x[idx], constants=obj.constants)
```
```suggestion
            getattr(obj, op)(v[idx], x[idx], constants=None)
```

YigitElma committed:
Update desc/objectives/objective_funs.py

YigitElma committed:
Update desc/objectives/objective_funs.py

YigitElma committed:
Update desc/objectives/objective_funs.py

YigitElma committed:
fix constants issue

YigitElma committed:
clean-up multidevice set_device and print backend, update some error messages

ddudt commented on desc/objectives/objective_funs.py:
```diff
@@ -465,20 +469,27 @@ def __init__(
-                f"({max(self._rank_per_objective)} in the array) "
-                f"is not equal to the number of ranks ({self.size}). There "
-                "should be at least 1 objective per rank.",
+                f"{max(self._rank_per_objective)} in the array) "
```
[![question](https://img.shields.io/badge/question-8B5CF6)](https://pullpo.io/cc?l=question) 
Why is the max value one more than the actual value? Say we have two ranks, then `rank_per_objective` should be `[0, 1]` and the max value is 1. Why would we print that the max value is 2?

YigitElma replied:
Since `self.size` is 2, printing both `max()` and `max()+1` seemed clearer. For example, let's say we have 3 ranks and `rank_per_objective` is [0, 1, 1], it will print
```
The maximum value of rank_per_objective (2, supplied as 1 in the 
array) is not equal to the number of ranks (3).
```
Your example won't trigger any errors.

YigitElma replied:
I am open to cleaner wording suggestions.

YigitElma replied:
I updated most of the error/warning messages

ddudt commented on docs/notebooks/tutorials/multi_device.ipynb:
```diff
@@ -677,9 +704,49 @@
-    "**Note: Most clusters have multiple GPUs connected to each node, so before using multiple nodes, use all the GPUs available to that node. Multi-node communication is significantly slower and your script will be easier to write properly.**\n",
+    "**Note: Most clusters have multiple GPUs connected to each node, so before using multiple nodes, use all the GPUs available to that node. Multi-node communication is slower and your script will be easier to write properly.**\n",
+    "\n",
+    "Note: You should have at least 6 objectives, so at least 1 objective per device. If you want to run multiple objectives on the same device, you can specify the ``rank_per_objective`` in the `ObjectiveFunction` keywords. By default, the initializer will assign different ranks for each sub-objective, in above example it defaults to `np.arange(len(objs))`. An example optimization setup where ea ...
```
[![question](https://img.shields.io/badge/question-8B5CF6)](https://pullpo.io/cc?l=question) 
Don't we allow the using fewer ranks than the total available? For example, if you have a node with 4 GPUs I thought you could run an optimization with only 3 objectives using 3 ranks on only 3 of the 4 devices. I thought this would throw a warning but still run correctly, with the extra GPU being unused.

YigitElma replied:
Yeah, you are correct, this is only a warning. I updated the explanation. The reason for 6 objectives is the number of ranks (above I say there will be 6 ranks) not number of devices

YigitElma committed:
update error/warning messages, update notebook

ddudt committed:
Merge branch 'master' into yge/multigpu

ddudt committed:
Merge branch 'master' into yge/multigpu

YigitElma committed:
update the message in test

daniel-dudt committed:
Merge branch 'master' into yge/multigpu

YigitElma committed:
Merge branch 'master' into yge/multigpu

YigitElma commented on docs/notebooks/tutorials/multi_device.ipynb:
add a note that GPU 0 is filled with null-space and all other matrices, and does all the heavy linear algebra.

YigitElma committed:
add the rank to _OBjective and construct rank_per_objective from it

YigitElma committed:
add rank to every objective

YigitElma committed:
update tests

YigitElma committed:
Merge branch 'master' into yge/multigpu

YigitElma committed:
default rank to None, assign np.arange(len(obj)) if none given, update tests, tutorials

YigitElma commented:
If you have been using this branch, I made a recent change. Now, instead of giving rank_per_objective to ObjectiveFunction, you should give it to each sub-objective. Default behavior of not having any rank is the same.

YigitElma commented:
I can finalize this PR for parallel constraints in Proximal too after #2239
